pkg/cmd/cli/cmd/idle.go

```diff
@@ -0,0 +1,539 @@
+		Short:   "Idle scalable resources",
+		Long:    idleLong,
+		Example: fmt.Sprintf(idleExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(RunIdle(f, out, cmd, inputFile, loadByService, dryRun))
+		},
+	}
+
+	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "If true, only print the annotations that would be written, without annotating or idling the relevant objects")
+	cmd.Flags().StringVarP(&inputFile, "filename", "f", "", "file containing list of scalable objects to idle in 'resource/name' format")
+	cmd.Flags().BoolVar(&loadByService, "by-service", loadByService, "If true, the input list will be treated as a list of services whose associated RCs and DCs should be idle, instead of being a list of RCs and DCs to idle")
+	cmd.MarkFlagRequired("filename")
+
+	return cmd
+}
+
+// getPodTemplateSpecNode finds the PodTemplateSpec node for a given RC node
+func getRCPodTemplateSpecNode(g osgraph.Graph, node graph.Node) *kubegraph.PodTemplateSpecNode {
+	rcSpecNodes := g.SuccessorNodesByNodeAndEdgeKind(node, kubegraph.ReplicationControllerSpecNodeKind, osgraph.ContainsEdgeKind)
+	if len(rcSpecNodes) != 1 {
+		panic(fmt.Sprintf("expected ReplicationControllerNode/DeploymentConfigNode %s to have a single ReplicationControllerSpecNode contained within it", g.GraphDescriber.Name(node)))
+	}
+
+	return getPodTemplateSpecNode(g, rcSpecNodes[0])
+}
+
+// getPodTemplateSpecNode finds the PodTemplateSpec node immediately contained in the given node
+func getPodTemplateSpecNode(g osgraph.Graph, node graph.Node) *kubegraph.PodTemplateSpecNode {
+	podTemplateSpecNodes := g.SuccessorNodesByNodeAndEdgeKind(node, kubegraph.PodTemplateSpecNodeKind, osgraph.ContainsEdgeKind)
+
+	if len(podTemplateSpecNodes) == 0 {
+		return nil
+	} else if len(podTemplateSpecNodes) > 1 {
+		panic(fmt.Sprintf("expected node %s to have no more than one PodTemplateSpecNode contained within it", g.GraphDescriber.Name(node)))
+	}
+
+	return podTemplateSpecNodes[0].(*kubegraph.PodTemplateSpecNode)
+}
+
+// getEndpointsNode finds the Endpoints node associated with a given Service node
+func getEndpointsNode(g osgraph.Graph, node osgraph.Node) *kubegraph.EndpointsNode {
+	endpointsNodes := g.SuccessorNodesByNodeAndEdgeKind(node, kubegraph.EndpointsNodeKind, kubeedges.HasEndpointsEdgeKind)
+	if len(endpointsNodes) != 1 {
+		panic(fmt.Sprintf("expected ServiceNode %s to have a single EndpointsNode corresponding to it", g.GraphDescriber.Name(node)))
+	}
+
+	return endpointsNodes[0].(*kubegraph.EndpointsNode)
+}
+
+// scanIdlableResourcesFromFile scans a file for either a list of resource names (when a default resource is provided),
+// or for a list of resources as resource/name (when no default resource is provided).  It returns a map from
+// resolved resource group-kind to a set of resource names.
+func scanIdlableResourcesFromFile(filename string, f *clientcmd.Factory, defaultResource string) (map[unversioned.GroupKind]map[string]bool, error) {
```

Comment:
>  The UI implementation is going to be really rough. Who is intended to call this? Admins?

Yeah, as far as I knew, the intended audience for the idling command was admins.  Is this getting a full equivalent in he web UI?  I was under the impression that since this was an admin command, mainly, that would not be the case.

> It seems like it would be easier to implement this by starting with the endpoints list

So, there is a mode which loads from endpoints instead, but it also currently uses the graph library (you can either start with a list of RCs and DCs to idle and find the services, or have a list of services to idle and find the DCs and RCs).

> That makes this command basically O(Endpoints + Controllers), and is completely generic (doesn't require the graph library). The inputs are either "all endpoints", or "list of endpoints".

So, it seems like that list of steps is kind of rewriting something similar to graph library, but without using the graph library.  What do we gain by dropping the graph library here?  Is the concern not being able to upstream as easily?  Is it not requiring extra calls (I already restrict the graph library to only load services, DCs, RCs, and endpoints, so it's not like we're loading the entire world here currently)?